pkg/apis/resource/types.go

```diff
@@ -0,0 +1,449 @@
+	// adding that information when it already allocated the resource.
+	//
+	// A driver must add entries here for all its pending claims, even if
+	// the ResourceSchedulingStatus.UnsuitabeNodes field is empty, because
+	// the scheduler may decide to wait with selecting a node until it has
+	// information from all drivers.
+	ResourceClaims []ResourceClaimSchedulingStatus
+
+	// If there ever is a need to support other kinds of resources
+	// than ResourceClaim, then new fields could get added here
+	// for those other resources.
+}
+
+// ResourceClaimSchedulingStatus contains information about one particular
+// claim while scheduling a pod.
+type ResourceClaimSchedulingStatus struct {
+	// PodClaimName matches the PodClaim.Name field.
+	PodClaimName string
+
+	// UnsuitableNodes lists nodes that the claim cannot be allocated for.
+	// Nodes listed here will be ignored by the scheduler when selecting a
+	// node for a Pod. All other nodes are potential candidates, either
+	// because no information is available yet or because allocation might
+	// succeed.
+	//
+	// A change of the PodSchedulingSpec.PotentialNodes field and/or a failed
+	// allocation attempt trigger an update of this field: the driver
+	// then checks all nodes listed in PotentialNodes and UnsuitableNodes
+	// and updates UnsuitableNodes.
+	//
+	// It must include the prior UnsuitableNodes in this check because the
+	// scheduler will not list those again in PotentialNodes but they might
+	// still be unsuitable.
+	//
+	// This can change, so the driver also must refresh this information
+	// periodically and/or after changing resource allocation for some
+	// other ResourceClaim until a node gets selected by the scheduler.
+	//
+	// The size of this field is limited to 128, the same as for
+	// PodSchedulingSpec.PotentialNodes.
+	UnsuitableNodes []string
+}
+
+// PodSchedulingNodeListMaxSize defines the maximum number of entries in the
+// node lists that are stored in PodScheduling objects. This limit is part
+// of the API.
+const PodSchedulingNodeListMaxSize = 128
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// PodSchedulingList is a collection of pod scheduling objects.
+type PodSchedulingList struct {
+	metav1.TypeMeta
+	// Standard list metadata
+	metav1.ListMeta
+
+	// Items is the list of resource claims.
+	Items []PodScheduling
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ResourceClass is used by administrators to influence how resources
+// are allocated.
+//
+// This is an alpha type and requires enabling the DynamicResourceAllocation
+// feature gate.
+type ResourceClass struct {
+	metav1.TypeMeta
+	// Standard object metadata
+	metav1.ObjectMeta
+
+	// DriverName determines which resource driver is to be used for
+	// allocation of a ResourceClaim that uses this class.
+	//
+	// Resource drivers have a unique name in reverse domain order
+	// (acme.example.com).
```

Comment:
```suggestion
	// DriverName defines the name of the dynamic resource driver that is used for
	// allocation of a ResourceClaim that uses this class.
	//
	// Resource drivers have a unique name in reverse domain order
	// (acme.example.com).
```

Aside: why reverse domain order? The rest of Kubernetes usually derives names from domain names in forward order.

Reply:
Copied.

Reverse domain order is used for consistency with CSI driver names.